File: apps/backend/app/schemas/attachment.py
"""
附件管理Schema
"""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from enum import Enum


class AttachmentStorageTypeEnum(str, Enum):
    """附件存储类型枚举"""
    LOCAL = "local"
    OSS = "oss"
    DATABASE = "database"


class AttachmentStatusEnum(str, Enum):
    """附件状态枚举"""
    UPLOADING = "uploading"
    ACTIVE = "active"
    ARCHIVED = "archived"
    DELETED = "deleted"


class AttachmentResponse(BaseModel):
    """附件响应"""
    id: int
    file_name: str
    original_name: str
    file_path: str
    file_size: int
    file_type: Optional[str] = None
    file_extension: Optional[str] = None
    file_md5: Optional[str] = None
    storage_type: AttachmentStorageTypeEnum
    entity_type: Optional[str] = None
    entity_id: Optional[int] = None
    category_id: Optional[int] = None
    uploaded_by: Optional[int] = None
    description: Optional[str] = None
    download_count: int
    status: AttachmentStatusEnum
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AttachmentUploadResponse(BaseModel):
    """附件上传响应"""
    attachment_id: int
    file_name: str
    file_size: int
    file_type: Optional[str] = None
    upload_url: Optional[str] = None
    message: str


class AttachmentDownloadUrl(BaseModel):
    """附件下载URL"""
    download_url: str
    expires_in: int  # 过期时间（秒）


class AttachmentQuery(BaseModel):
    """附件查询参数"""
    entity_type: Optional[str] = None
    entity_id: Optional[int] = None
    category_id: Optional[int] = None
    uploaded_by: Optional[int] = None
    status: Optional[AttachmentStatusEnum] = None
    file_type: Optional[str] = None  # 文件类型筛选（image/*, application/pdf等）
    keyword: Optional[str] = None  # 搜索关键词
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


class AttachmentCategoryResponse(BaseModel):
    """附件分类响应"""
    id: int
    name: str
    code: str
    description: Optional[str] = None
    allowed_types: Optional[str] = None
    max_size: Optional[int] = None
    is_active: bool
    sort_order: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AttachmentCategoryCreate(BaseModel):
    """创建附件分类"""
    name: str = Field(..., max_length=100)
    code: str = Field(..., max_length=50)
    description: Optional[str] = None
    allowed_types: Optional[str] = None  # 如：image/*,application/pdf,.doc,.docx
    max_size: Optional[int] = None  # 字节
    sort_order: int = Field(default=0)


class AttachmentCategoryUpdate(BaseModel):
    """更新附件分类"""
    name: Optional[str] = Field(None, max_length=100)
    description: Optional[str] = None
    allowed_types: Optional[str] = None
    max_size: Optional[int] = None
    is_active: Optional[bool] = None
    sort_order: Optional[int] = None

File: apps/frontend/src/types/attachment.ts
/**
 * 附件管理类型定义
 * @description 附件相关的数据结构，用于前后端数据交互
 */

/**
 * 附件存储类型枚举
 */
export enum AttachmentStorageType {
  LOCAL = "local",
  OSS = "oss",
  DATABASE = "database",
}

/**
 * 附件状态枚举
 */
export enum AttachmentStatus {
  UPLOADING = "uploading",
  ACTIVE = "active",
  ARCHIVED = "archived",
  DELETED = "deleted",
}

/**
 * 附件响应数据结构
 */
export interface Attachment {
  id: number;
  file_name: string;
  original_name: string;
  file_path: string;
  file_size: number;
  file_type?: string;
  file_extension?: string;
  file_md5?: string;
  storage_type: AttachmentStorageType;
  entity_type?: string;
  entity_id?: number;
  category_id?: number;
  uploaded_by?: number;
  description?: string;
  download_count: number;
  status: AttachmentStatus;
  created_at: string;
  updated_at: string;
}

/**
 * 附件上传响应
 */
export interface AttachmentUploadResponse {
  attachment_id: number;
  file_name: string;
  file_size: number;
  file_type?: string;
  upload_url?: string;
  message: string;
}

/**
 * 附件上传请求
 */
export interface AttachmentUploadRequest {
  file: File;
  entity_type?: string;
  entity_id?: number;
  category_id?: number;
  description?: string;
}

/**
 * 附件查询参数
 */
export interface AttachmentQuery {
  entity_type?: string;
  entity_id?: number;
  category_id?: number;
  uploaded_by?: number;
  status?: AttachmentStatus;
  file_type?: string;
  keyword?: string;
  page?: number;
  page_size?: number;
}

/**
 * 附件列表响应
 */
export interface AttachmentListResponse {
  items: Attachment[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 附件分类
 */
export interface AttachmentCategory {
  id: number;
  name: string;
  code: string;
  description?: string;
  allowed_types?: string;
  max_size?: number;
  is_active: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

/**
 * 附件分类创建请求
 */
export interface AttachmentCategoryCreate {
  name: string;
  code: string;
  description?: string;
  allowed_types?: string;
  max_size?: number;
  sort_order?: number;
}

/**
 * 附件分类更新请求
 */
export interface AttachmentCategoryUpdate {
  name?: string;
  description?: string;
  allowed_types?: string;
  max_size?: number;
  is_active?: boolean;
  sort_order?: number;
}

/**
 * 上传进度回调
 */
export type UploadProgressCallback = (progress: number) => void;
